rfcs/20240722-group-query-attention/README.md

```diff
@@ -0,0 +1,215 @@
+3. Performance 5D matmul between (N, G, H / G, S, D) and (N, G, 1, S, D). The
+   third dimension will be broadcasted from 1 to `H / G` automatically per the
+   broadcasting rule of MatMul operation.
+4. Similar reshape and broadcasting can also be applied to the dot-product of
+   Value.
+
+Here is the diagram:
+![option2](option2.png)
+
+Pros:
+
+1. No change to API.
+2. The pattern looks simpler than the pattern in option 1.
+
+Cons:
+
+1. The pattern is less intuitive from GQA definition.
+2. The pattern cannot be used to optimize a framework graph directly. Frameworks
+   will have to implement GQA fusion by themselves and leverage this option to
+   optimized the fused GQA.
+
+### Option 3
+
+We can extend the MatMul broadcasting rules to support group broadcast.
+Currently, MatMul operation supports below broadcasting rules.
+
+For single 2D matrix multiplication:
+
+| Matrix A | Matrix B | Matrix C = A x B |
+| --       | --       | --               |
+| M x K    | K x N    | M x N            |
+
+For batched matrix multiplications:
+
+| Matrix A   | Matrix B   | Matrix C = A x B |
+| --         | --         | --               |
+| B x M x K  | B x K x N  | B x M x N        |
+| B x M x K  | 1 x K x N  | B x M x N        |
+| 1 x M x K  | B x K x N  | B x M x N        |
+
+This can be extended to multiple batch dimensions matrix multiplications:
+
+| Matrix A              | Matrix B              | Matrix C = A x B      |
+| --                    | --                    | --                    |
+| B1 x 1 x B3 x M x K   | B1 x B2 x 1 x M x K   | B1 x B2 x B3 x M x N  |
+
+This RFC proposes to further extend it to support group broadcast:
+
+| Matrix A              | Matrix B                | Matrix C = A x B      |
+| --                    | --                      | --                    |
+| B1 x 1 x B3 x M x K   | 1 x B2 x (B3/c) x K x N | B1 x B2 x B3 x M x N  |
+
+where, c is a factor of B3.
+
+This rule looks uncommon and is not supported by the typical broadcasting rules
+(see broadcasting in
+[ONNX](https://github.com/onnx/onnx/blob/main/docs/Broadcasting.md) and
+[NumPy](https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules).),
+but actually it's added to the MatMul operation of cuDNN in order to support
+GQA.
+
+Pros.
+
+1. We will have the most unified and simple pattern for SDPA, GQA, and MQA.
+2. The same support methodology with cuDNN API.
+
+Cons.
+
+1. It complicates the semantics of MatMul operation. Previously incompatible
+   batch dimension will be treated as an error. But now we need to further check
+   if they can be properly group broadcasted. Even cuDNN explicitly document
+   that the new broadcasting rule is only supported by the fused attention.
+2. Same as option 2, still the pattern cannot be used to optimize a framework
+   graph directly. Frameworks will have to implement GQA fusion by themselves
+   and leverage this option to optimized the fused GQA.
```

Comment:
Another con here maybe that we rely on oneDNN matmul primitive kernels for reference implementation and testing in benchdnn which do not support the new broadcasting rule. Extending the broadcast semantics on graph side will also request additional effort for reference implementation and testing.

Reply:
Added, thanks.